share/core.go

```diff
@@ -0,0 +1,315 @@
+// Package share implements different secret sharing mechanisms, including
+// simple Shamir secret sharing, verifiable secret sharing (vss), and public
+// verifiable secret sharing (pvss).
+package share
+
+import (
+	"crypto/cipher"
+	"errors"
+
+	"github.com/dedis/crypto/abstract"
+)
+
+// This file provides the core functionality for secret sharing including
+// Shamir's scheme and public commitment polynomials.
+
+// Some error definitions
+var errorGroups = errors.New("non-matching groups")
+var errorCoeffs = errors.New("different number of coefficients")
+
+// PriShare represents a private share.
+type PriShare struct {
+	I int             // Index of the private share
+	V abstract.Scalar // Value of the private share
+}
+
+// PriPoly represents a secret sharing polynomial.
+type PriPoly struct {
+	g      abstract.Group    // Cryptographic group
+	coeffs []abstract.Scalar // Coefficients of the polynomial
+}
+
+// NewPriPoly creates a new secret sharing polynomial for the cryptographic
+// group g, the secret sharing threshold t, and the secret to be shared s.
+func NewPriPoly(g abstract.Group, t int, s abstract.Scalar, rand cipher.Stream) *PriPoly {
+	coeffs := make([]abstract.Scalar, t)
+	coeffs[0] = s
+	if coeffs[0] == nil {
+		coeffs[0] = g.Scalar().Pick(rand)
+	}
+	for i := 1; i < t; i++ {
+		coeffs[i] = g.Scalar().Pick(rand)
+	}
+	return &PriPoly{g, coeffs}
+}
+
+// Threshold returns the secret sharing threshold.
+func (p *PriPoly) Threshold() int {
+	return len(p.coeffs)
+}
+
+// GetSecret returns the shared secret p(0), i.e., the constant term of the polynomial.
+func (p *PriPoly) GetSecret() abstract.Scalar {
+	return p.coeffs[0]
+}
+
+// Eval computes the private share v = p(i).
+func (p *PriPoly) Eval(i int) *PriShare {
+	xi := p.g.Scalar().SetInt64(1 + int64(i)) // x-coordinate of this share
+	v := p.g.Scalar().Zero()
+	for j := p.Threshold() - 1; j >= 0; j-- {
+		v.Mul(v, xi)
+		v.Add(v, p.coeffs[j])
+	}
+	return &PriShare{i, v}
+}
+
+// Shares creates a list of n private shares p(1),...,p(n).
+func (p *PriPoly) Shares(n int) []*PriShare {
+	shares := make([]*PriShare, n)
+	for i := 0; i < n; i++ {
+		shares[i] = p.Eval(i)
+	}
+	return shares
+}
+
+// Add computes the component-wise sum of the polynomials p and q and returns it
+// as a new polynomial.
+func (p *PriPoly) Add(q *PriPoly) (*PriPoly, error) {
+
+	if p.g != q.g {
+		return nil, errorGroups
+	}
+
+	if p.Threshold() != q.Threshold() {
+		return nil, errorCoeffs
+	}
+
+	t := p.Threshold()
+	coeffs := make([]abstract.Scalar, t)
+	for i := 0; i < t; i++ {
+		coeffs[i] = p.g.Scalar().Add(p.coeffs[i], q.coeffs[i])
+	}
+
+	return &PriPoly{p.g, coeffs}, nil
+}
+
+// Equal checks equality of two secret sharing polynomials p and q.
```

Comment:
Not sure I understand what you mean, because checking all coefficients (or commits in this particular case) is exactly what I did. Or do you suggest I use `P.Equal(Q)` for points/commits `P` and `Q`?